### PR TITLE
SSN Survey Functionality Updates

### PR DESCRIPTION
**Pull request by anthonypetersen**
This PR addresses the following Issue:
* https://github.com/episphere/connect/issues/712
-----
## Background Details
* The SSN Survey is being moved outside of Quest and into an on-screen Frame within the To-Do List of the PWA. Because of this the data coming into the API will be slightly different and this also gives us a chance to separate the logic for submitting SSN details from other survey responses.
-----
## Technical Changes
* Added new option to `connectApp` API
* Removed `decryptionAsymmetric()` function from `encrypt.js`
* Added new SSN constants to `fieldToConceptIdMapping.js`
* Refactored and moved SSN submission logic to a separate function within `submission.js`
* Code cleanup in the following files: `index.js`, `firestore.js`, `shared.js`
 

**NOTE: Should be moved along tiers with [PWA PR](https://github.com/episphere/connectApp/pull/698)**

### CODE REVIEW

File: utils/submission.js [hunk 1]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
     return submit(res, data, uid);
 }
 
+const submitSocial = async (req, res, uid) => {
+    if(req.method !== 'POST') {
```
Comment (Gbarra9) [inline utils/submission.js:131]:
```suggestion
    if (req.method !== 'POST') {
```

File: utils/submission.js [hunk 2]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
...
+        return res.status(405).json(getResponseJSON('Only POST requests are accepted!', 405));
+    }
+
+    const data = req.body;
+    if(Object.keys(data).length <= 0){
```
Comment (Gbarra9) [inline utils/submission.js:136]:
```suggestion
    if (Object.keys(data).length <= 0){
```

File: utils/submission.js [hunk 3]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
...
+        return res.status(400).json(getResponseJSON('Bad request!', 400));
+    }
+
+    if (!data['ssnNineDigits'] && !data['ssnFourDigits']) {
+        return res.status(400).json(getResponseJSON('Bad request!', 400));
+    }
+
+    // SSN 9 digits
+    if(data['ssnNineDigits']) {
```
Comment (Gbarra9) [inline utils/submission.js:145]:
```suggestion
    if (data['ssnNineDigits']) {
```

File: utils/submission.js [hunk 4]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
...
+        const ssn = data['ssnNineDigits'];
+        const ssnObj = {};
+
+        ssnObj[fieldMapping.ssnFullValue] = await encryptAsymmetric(ssn.replace(/-/g, ''));
+        ssnObj[fieldMapping.ssnQcStatus] = fieldMapping.ssnNoCheckRan;
+        ssnObj['uid'] = uid;
+        ssnObj['token'] = await getTokenForParticipant(uid);
+
+        let ssnResponse = storeSSN(ssnObj);
+        if(ssnResponse instanceof Error){
```
Comment (Gbarra9) [inline utils/submission.js:158]:
```suggestion
        if (ssnResponse instanceof Error){
```

File: utils/submission.js [hunk 5]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
...
+        }
+
+        data[fieldMapping.ssnFullGiven] = 353358909;
+        data[fieldMapping.ssnFullGivenTime] = new Date().toISOString();
+
+        delete data['ssnNineDigits'];
+    }
+
+    // SSN last 4 digits
+    if(data['ssnFourDigits']) {
```
Comment (Gbarra9) [inline utils/submission.js:169]:
```suggestion
    if (data['ssnFourDigits']) { 
```

File: utils/submission.js [hunk 6]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
...
+        const ssn = data['ssnFourDigits'];
+        const ssnObj = {};
+
+        ssnObj[fieldMapping.ssnPartialValue] = await encryptAsymmetric(ssn.replace(/-/g, ''));
+        ssnObj[fieldMapping.ssnQcStatus] = fieldMapping.ssnNoCheckRan;
+        ssnObj['uid'] = uid;
+        ssnObj['token'] = await getTokenForParticipant(uid);
+
+        let ssnResponse = storeSSN(ssnObj);
+        if(ssnResponse instanceof Error){
```
Comment (Gbarra9) [inline utils/submission.js:182]:
```suggestion
        if (ssnResponse instanceof Error){
```

File: utils/submission.js [hunk 7]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
...
+        delete data['ssnFourDigits'];
+    }
+
+    data[fieldMapping.ssnCompleteTs] = new Date().toISOString();
+    data[fieldMapping.ssnStatusFlag] = 231311385;
+
+    const { updateResponse } = require('./firestore');
+    const response = await updateResponse(data, uid);
+
+    if(response instanceof Error){
```
Comment (Gbarra9) [inline utils/submission.js:198]:
```suggestion
    if (response instanceof Error){
```

File: utils/submission.js [hunk 8]
```diff
@@ -166,6 +127,83 @@ const recruitSubmit = async (req, res, uid) => {
...
+        return res.status(500).json(getResponseJSON(response.message, 500));
+    }
+    if(!response) {
```
Comment (Gbarra9) [inline utils/submission.js:201]:
```suggestion
    if (!response) {
```

**Review by Gbarra9 — APPROVED** [review]

**Review by bransteitterbr — APPROVED** [review]

**Comment by bransteitterbr** [comment]
looks good

File: utils/submission.js [hunk 9]
```diff
@@ -166,6 +127,80 @@ const recruitSubmit = async (req, res, uid) => {
...
+    delete data['ssnNineDigits'];
+    delete data['ssnFourDigits'];
+
+    if (Object.keys(data).length <= 0){
+        return res.status(400).json(getResponseJSON('Bad request!', 400));
+    }
+
+    if (!ssnNineDigits && !ssnFourDigits) {
+        return res.status(400).json(getResponseJSON('Bad request!', 400));
+    }
```
Comment (we-ai) [inline utils/submission.js:149]: Can we combine these 2 condition checks with  OR  ?

**Review by we-ai — APPROVED** [review]

**Pull request merged**